packages/program/src/functions/base-invocation-scope.ts

```diff
@@ -216,22 +223,53 @@ export class BaseInvocationScope<TReturn = any> {
   }
 
   /**
-   * Gets the transaction cost ny dry running the transaction.
+   * Gets the transaction cost for dry running the transaction.
    *
    * @param options - Optional transaction cost options.
    * @returns The transaction cost details.
    */
   async getTransactionCost() {
     const provider = this.getProvider();
+    const baseAssetId = provider.getBaseAssetId();
+    const request = clone(await this.getTransactionRequest());
+    const isScriptTransaction = request.type === TransactionType.Script;
+
+    // Fund with fake UTXOs to avoid not enough funds error
+    // Getting coin quantities from amounts being transferred
+    const coinOutputsQuantities = request.getCoinOutputsQuantities();
+    // Combining coin quantities from amounts being transferred and forwarding to contracts
+    const quantitiesToContract = this.getRequiredCoins();
+    const requiredQuantities = mergeQuantities(coinOutputsQuantities, quantitiesToContract);
+    // Funding transaction with fake utxos
+    request.fundWithFakeUtxos(requiredQuantities, baseAssetId, this.program.account?.address);
+
+    /**
+     * Estimate predicates gasUsed
+     */
+    // Remove gasLimit to avoid gasLimit when estimating predicates
+    if (isScriptTransaction) {
+      request.gasLimit = bn(0);
+    }
+
+    /**
+     * The fake utxos added above can be from a predicate
+     * If the resources owner is a predicate,
+     * we need to populate the resources with the predicate's data
+     * so that predicate estimation can happen.
+     */
+    if (this.program.account && 'populateTransactionPredicateData' in this.program.account) {
+      (this.program.account as unknown as Predicate<[]>).populateTransactionPredicateData(request);
+    }
```

Comment:
If the `Account.getTransactionCost` is already doing the same thing we have here, why do we need this code here?

Reply:
The `getTransactionCost` here can be called by any of the `dryRun` methods so if the account is not defined, we can simply generate a new one using the provider. This method is only used for estimation purposes.